[angular-xmcloud][xmcloud-proxy] Personalization support

art-alexeyenko opened the pull request:
- [x] This PR follows the [Contribution Guide](https://github.com/Sitecore/jss/blob/dev/CONTRIBUTING.md)
- [x] Changelog updated
- [ ] Upgrade guide entry added

## Description / Motivation
Adds personalization and A/B testing support into angular-xmcloud setup.
Personalization works through xmcloud-proxy and should allow for personalization support for react and vue in future.

## Testing Details
- [x] Unit Test Added
- [x] Manual Test/Other (Please elaborate)

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

yavorsk commented on packages/create-sitecore-jss/src/templates/node-xmcloud-proxy/src/personalize.ts:
```diff
@@ -0,0 +1,30 @@
+import { responseInterceptor } from 'http-proxy-middleware';
+import { Plugin } from 'http-proxy-middleware/dist/types';
+import { IncomingMessageWithBody } from './types';
+export const personalizeHelper = new PersonalizeHelper(personalizeConfig);
```
```suggestion

export const personalizeHelper = new PersonalizeHelper(personalizeConfig);
```

yavorsk replied:
can add blank line after the imports

yavorsk commented on packages/sitecore-jss-angular/src/public_api.ts:
```diff
@@ -80,10 +81,20 @@ export {
+  GraphQLPersonalizeService,
+  GraphQLPersonalizeServiceConfig,
+  PersonalizeInfo,
+  CdpHelper,
```
the import of CdpHelper at the end of this file should be removed to fix the build

yavorsk reviewed (approved):
Great work! :)

could not find anything significant, just few minor things - one of which to fix the build - otherwise approve from me!

yavorsk commented:
One more thing - do we need some migration instructions to the upgrade guide?

illiakovalenko commented on packages/create-sitecore-jss/src/templates/node-xmcloud-proxy/src/personalize.ts:
```diff
@@ -0,0 +1,30 @@
+    responseInterceptor(async (responseBuffer, _, req, res) => {
+      let responseText = responseBuffer.toString('utf8');
+      const payload = JSON.stringify((req as IncomingMessageWithBody).body);
+      if (payload.toLowerCase().includes('jsslayoutquery')) {
```
We need to introduce a variable GRAPHQL_LAYOUT_QUERY_NAME in "sitecore-jss". It should be referenced value instead of hardcoded

illiakovalenko commented on packages/create-sitecore-jss/src/templates/node-xmcloud-proxy/src/index.ts:
```diff
@@ -165,11 +174,17 @@ server.use(async (req, res) => {
     );
-
+    // for SSR loading routing, personalization is performed by modifying layoutData directly
+    const personalizedLayoutData = await personalizeHelper.personalizeLayoutData(
```
We are still not handling the case when "apiKey" is used instead of "contextId"
otherwise the personalization will fail. This is what we discussed yesterday

art-alexeyenko replied:
I adjusted the setup on angular app side for this.

illiakovalenko replied:
So, we don't have any layout requests targeted against proxy (when performing client side navigation)
however, we are still trying to "personalize" when user opens the page first time (SSR)

illiakovalenko commented on packages/sitecore-jss-proxy/tsconfig.json:
```diff
@@ -5,8 +5,9 @@
     "module": "commonjs",
     "outDir": "dist/cjs",
     "strictBindCallApply": false,
+    "skipLibCheck": true,
```
Do we have some library that is failing now? cloudsdk 0.4?

art-alexeyenko replied:
yes, it's cloudsdk

illiakovalenko replied:
I agree to add this option, but may be there is a problem in cloudsdk 0.4? what is the problem, did you report it (created a bug)?

art-alexeyenko replied:
Good idea, I will add a bug.
Otherwise this option just brings proxy package in line with other sitecore-jss packages.

illiakovalenko commented on packages/sitecore-jss-proxy/src/personalize/PersonalizeHelper.ts:
```diff
@@ -0,0 +1,310 @@
+
+    const result = variantIds ? this.personalizeLayout(layoutData, variantIds) : layoutData;
+    debug.personalize('personalize layout end in %dms: %o', Date.now() - startTimestamp, {
+      headers: this.extractDebugHeaders(req.headers),
```
Let's add variantIds as well

illiakovalenko commented on packages/sitecore-jss-proxy/src/personalize/PersonalizeHelper.ts:
```diff
@@ -0,0 +1,310 @@
+   * Init CloudSDK personalization on server side
+   * @param {IncomingMessage} request incoming nodejs request object
+   * @param {OutgoingMessage} response outgoing nodejs response object
+   * @param {string} [hostname] host for cookies. When not provided, host will be read from host header, and fallback to 'localhost' if that fails
```
```suggestion
   * @param {string} hostname host for cookies. When not provided, host will be read from host header, and fallback to 'localhost' if that fails
```

illiakovalenko commented on packages/sitecore-jss-proxy/src/personalize/PersonalizeHelper.ts:
```diff
@@ -0,0 +1,310 @@
+    language: string,
+    pathname: string
+  ): Promise<string[]> => {
+    // const startTimestamp = Date.now();
```
should it be removed?

illiakovalenko commented on packages/sitecore-jss-proxy/src/personalize/PersonalizeHelper.ts:
```diff
@@ -0,0 +1,310 @@
+
+  protected getExperienceParams(req: IncomingMessage): ExperienceParams {
+    // nodejs req.url does not have a hostname, we parse query string the old fashioned way
+    const rawQs = req.url?.split('?')[1] || '';
```
Suggest you to use more "safe" approach
```ts
const rawQs = req.url ? new URL(req.url).searchParams : '';
```

art-alexeyenko replied:
req.url does not have a hostname - we cannot transform it with `new URL` reliably.

illiakovalenko replied:
got it
makes sense now

illiakovalenko commented on packages/sitecore-jss-proxy/src/personalize/PersonalizeHelper.ts:
```diff
@@ -0,0 +1,310 @@
+  }
+
+  protected getLanguage(layoutData: LayoutServiceData): string {
+    return layoutData.sitecore?.context?.language || 'en';
```
I think we need to add another check for "defaultLanguage"
XMC Proxy currently accepts "defaultLanguage" as a part of the "serverBundle", however in this case here it's ignored and our default value is prioritized

illiakovalenko commented on packages/sitecore-jss-proxy/src/personalize/PersonalizeHelper.ts:
```diff
@@ -0,0 +1,310 @@
+  };
+
+  protected personalizeLayout(layoutData: LayoutServiceData, variantIds: string[]) {
+    if (!layoutData.sitecore?.route) {
```
Can we add this check at the very start of the flow? In "personalizeLayoutData" method.
So, we will not be doing anything

illiakovalenko commented on packages/sitecore-jss-proxy/src/personalize/PersonalizeHelper.ts:
```diff
@@ -0,0 +1,310 @@
+      personalizeData.variantId,
+      personalizeData.componentVariantIds
+    );
+    layoutData.sitecore.route.placeholders = personalizedPlaceholders;
```
I think you don't need to do that, since "personalizeLayout" updates "layoutData" structure directly

art-alexeyenko replied:
I have found this approach to be unlreliable when running unit tests. Also, directly assigning the result is more in my opinion.

illiakovalenko replied:
the thing that even if you do that, `layoutData` is already updated by the "personalizeLayout" function. You are assigning the same value. Why we can't do exactly like we are doing in next.js?

art-alexeyenko commented on packages/create-sitecore-jss/src/templates/node-xmcloud-proxy/src/config.ts:
```diff
@@ -23,3 +28,33 @@ export const config: Config = {
+  // Configuration for your Sitecore CDP endpoint
+  // Edge URL and ID can be taken from proxy env, or the base SPA app
+  cdpConfig: {
+    sitecoreEdgeUrl: process.env.SITECORE_EDGE_URL || fallbackEdgeEndpoint,
+    sitecoreEdgeContextId: process.env.SITECORE_EDGE_CONTEXT_ID || fallbackEdgeId || '',
```
I respectfully disagree with this, but it is something we should discuss further as a separate talk.
For now this is a stopgap measure until we figure out the conflict between SPA and proxy env variables.

illiakovalenko replied:
so, correct - this thing is early to add, until we introduce support for clientside env variables
otherwise you can define some variables in proxy and in the browser you will pull values from SPA app env variables, this is the mismatch right now

art-alexeyenko commented:
@illiakovalenko @yavorsk Addressed your comments, please take a look.
In regards to upgrade guide, I added some light instructions. Otherwise, there's not much to instruct, right? Xmcloud proxy is a new product, users would simply need to make the switch and start using it.

illiakovalenko commented on docs/upgrades/unreleased.md:
````diff
@@ -129,6 +129,21 @@
+        "@sitecore-cloudsdk/core": "^0.4.0",
+        "@sitecore-cloudsdk/events": "^0.4.0",
+    ```
+* Add an extra export in `server.exports.ts`:
````
We already have a section below that asks to add this file, can you merge this change under existing section?

illiakovalenko commented on packages/sitecore-jss/src/constants.ts:
```diff
@@ -21,3 +21,5 @@ export const siteNameError = 'The siteName cannot be empty';
 
 export const HIDDEN_RENDERING_NAME = 'Hidden Rendering';
+
+export const GRAPHQL_LAYOUT_QUERY_NAME = 'JssLayoutQuery';
```
since it's a part of "layout" submodule, can we add it there instead?

illiakovalenko commented:
@art-alexeyenko Can you, please, check code coverage as well? use "npm run coverage", I've seen that code snippets are not covered

art-alexeyenko commented on packages/create-sitecore-jss/src/templates/node-xmcloud-proxy/src/index.ts:
```diff
@@ -125,6 +128,12 @@ server.use(
     changeOrigin: true,
+    selfHandleResponse: true,
+    on: {
+      proxyReq: fixRequestBody,
```
exactly right. Layout Service requests are POST requests. We don't want them to end prematurely while personalization logic is running.

art-alexeyenko replied:
And if you mean `fixRequestBody` then we definitely need this, as request modified by `express.json()` will hang otherwise

illiakovalenko replied:
Got it, thanks for explanation!

art-alexeyenko commented on packages/sitecore-jss-proxy/src/personalize/PersonalizeHelper.test.ts:
```diff
@@ -0,0 +1,658 @@
+
+        const { helper, initPersonalizeServer } = createHelper(props);
+
+        await helper.personalizeLayoutData(req, res, defaultLayoutData);
```
hostname is tested here, and it has no effect on the layout end result, so we don't test the layout output. irregularities of layout result will be tested in other parts

art-alexeyenko commented on packages/sitecore-jss-proxy/src/personalize/PersonalizeHelper.test.ts:
```diff
@@ -0,0 +1,658 @@
+    const getComponentFriendlyIdStub = sinon.stub();
+    const getPageFriendlyIdStub = sinon.stub();
+    const PersonalizeHelper = proxyquire('./PersonalizeHelper', {
+      '@sitecore-jss/sitecore-jss/personalize': {
```
The internal function is still used. Stub is only there to keep track of function calls, as `sinon.spy` does not capture the method call OOB. see the comment below this line.

art-alexeyenko replied:
So this is basically a passthu stub

illiakovalenko replied:
Got it, it still uses the original func, I missed that
this function name is quite similar in other places, lol

art-alexeyenko commented on packages/sitecore-jss/src/personalize/layout-personalizer.ts:
```diff
@@ -29,12 +29,12 @@ export function personalizeLayout(
-    return;
+  const placeholders = layout.sitecore.route?.placeholders || {};
+  if (Object.keys(placeholders).length === 0) {
+    return {};
```
I'd leave this in for the future. Without it the return type of the function is calculated to "PlaceholderData | undefined", it's clunky and there's no need to have undefined there.

illiakovalenko replied:
the thing that it's a breaking change, technically

illiakovalenko replied:
if someone had a check for "undefined" now it always returns some value

illiakovalenko commented on packages/sitecore-jss-proxy/src/personalize/PersonalizeHelper.test.ts:
```diff
@@ -0,0 +1,658 @@
+
+        const { helper, initPersonalizeServer } = createHelper();
+
+        await helper.personalizeLayoutData(req, res, defaultLayoutData);
```
closed

illiakovalenko commented on packages/create-sitecore-jss/src/templates/node-xmcloud-proxy/src/types.ts:
```diff
@@ -25,3 +28,10 @@ export interface Config {
+/**
+ * IncomingMessage type modified with exporess.json() call to include request body
+ */
+export type IncomingMessageWithBody = IncomingMessage & {
```
This one I close since we agreed to leave it as its

illiakovalenko reviewed (approved):
Approved in advance! Great progress, it looks cleaner now 👍 
See one comment below and revert the piece that produces breaking changes `personalize/layoutPersonalizer`
